src/SharpLearning.Backend.Cntk.Test/CntkRawMnistTest.cs

```diff
@@ -0,0 +1,585 @@
+﻿using System;
+using System.IO;
+using System.Linq;
+using System.IO.Compression;
+using System.Net;
+using CNTK;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+// To avoid name-clash with SharpLearning.Backend.DataType.
+using CntkDataType = CNTK.DataType;
+using System.Text;
+
+namespace SharpLearning.Backend.Cntk.Test
+{
+    [TestClass]
+    public class CntkRawMnistTest
+    {
+        const string DownloadPath = "MnistTest";
+
+        [TestMethod]
+        public void MnistTrainTest()
+        {
+            // Data set
+            var mnist = Mnist.Load(DownloadPath);
+            var trainingSampleCount = mnist.TrainImages.Length;
+            var dataShape = new int[] { 28, 28, 1 };
+            var numberOfClasses = 10;
+            var outputValidationError = false;
+
+            // Set global data and device types. 
+            var dataType = CntkDataType.Float;
+            var device = DeviceDescriptor.UseDefaultDevice();
+            Trace.WriteLine("Using device: " + device.Type);
+            // These are just hardcoded in the helper class.
+            CntkLayers.DataType = dataType;
+            CntkLayers.Device = device;
+
+            // Define small convolutional neural network.
+            var net = CntkLayers.Input(dataShape);
+
+            net = CntkLayers.Conv2D(net, 3, 3, 32);
+            net = CntkLayers.ActivationFunction(net, Activation.ReLU);
+
+            net = CntkLayers.Pool2D(net, 2, 2);
+
+            net = CntkLayers.Dense(net, 256);
+            net = CntkLayers.ActivationFunction(net, Activation.ReLU);
+
+            net = CntkLayers.Dropout(net, 0.5);
+
+            net = CntkLayers.Dense(net, numberOfClasses);
+            net = CntkLayers.SoftMax(net);
+
+            // Variables for setting features and targets while training the network.
+            var featureVariable = net.Arguments[0];
+            var targetVariable = Variable.InputVariable(new int[] { numberOfClasses }, dataType);
+
+            // Training loss and eval error. Eval error is only for reporting and not used for training.
+            var loss = CNTKLib.CrossEntropyWithSoftmax(net, targetVariable);
+            var evalError = CNTKLib.ClassificationError(net, targetVariable);
+
+            // Setup stochastic gradient descent learner
+            var learningRatePerSample = new TrainingParameterScheduleDouble(0.01, 1);
+            var parameterLearners = new List<Learner>() { Learner.SGDLearner(net.Parameters(), learningRatePerSample) };
```

Comment:
@mdabros this also means I think we should drop "var" usage here for CNTK types, so we can better see the types involved. This is not code for a library it is code for "learning" the shape of the API, with the abstractions introduced we lose sight of the goal here. Abstractions hide what we want to "learn" about.

Reply:
Yes, I think this is a good idea.